Add support for `tsh ssh` on Windows

timothyb89 opened the pull request:
This adds Windows session support to tsh, taking advantage of ANSI terminal support and VT emulation added in recent versions Windows 10. On supported Windows versions (Windows 10 1607+), `tsh ssh` should work as expected in `cmd.exe`, PowerShell, and the new Windows Terminal app.

This uses go-ansiterm's [`winterm` module](https://github.com/Azure/go-ansiterm/tree/master/winterm) to configure the terminal for raw input and output. As this isn't sufficient to capture all keystrokes, we additionally implement a custom terminal input library derived from the input handling component of Microsoft's [OpenSSH fork](https://github.com/PowerShell/openssh-portable/tree/latestw_all/contrib/win32/win32compat).

Fixes #5415

timothyb89 committed:
Add support for `tsh ssh` on Windows

This adds Windows session support to tsh, taking advantage of ANSI
terminal support and VT emulation added in recent versions Windows
10. On supported Windows versions (Windows 10 1809+ and Server
2019+), `tsh ssh` should work as expected in `cmd.exe`, PowerShell,
and the new Windows Terminal app.

timothyb89 commented:
A few things I'm specifically looking for feedback on:
 * There's quite a lot of duplicated code between `session.go` and `session_windows.go`. A couple identical functions could be hoisted into a `session_common.go` and shared, but most require small compatibility tweaks. I'd love to hear any thoughts for more elegant ways to refactor this.
 * I don't _think_ I've introduced any memory safety issues (at least, none that aren't also present in the imported C code) but would very much appreciate a 2nd opinion from someone with more C experience.
 * tsh makes assumptions about default usernames that are generally invalid on Windows. I partially fixed this by stripping the domain off the username string, but `tsh ssh` and `tsh join` still generally fail when run without an explicit remote username. Is there a more sensible username we could use on Windows?
 * I'm sure copyright headers on the imported/modified code need changes.

Also, a few known issues:
 * This further breaks `tsh ssh` in Cygwin/msys2 (which was at least partially functional before).
 * tsh still emits raw VT sequences in log messages which aren't processed. I think this may be best addressed in another PR.
 * There's no tests on any of this. I'm not sure we have the ability to even run them, let alone deal with Windows versions and GUI terminal requirements.

zmb3 commented on lib/client/session_windows.go:
```diff
@@ -20,23 +20,208 @@ package client
+		return fmt.Errorf(`syscalls required for tsh.exe ssh on Windows could not be found.
+Ensure Windows 10 1809 or later is installed.
+
+Missing syscalls: %s`, missing)
```
```suggestion
Missing syscalls: %s`, strings.Join(missing, ", "))
```

zmb3 commented on lib/client/session_windows.go:
```diff
@@ -20,23 +20,208 @@ package client
+Ensure Windows 10 1809 or later is installed.
+
+Missing syscalls: %s`, missing)
+	} else {
```
Remove the else?

zmb3 commented on lib/client/session_windows.go:
```diff
@@ -20,23 +20,208 @@ package client
+	}
+
+	return func() {
+		err = winterm.SetConsoleMode(uintptr(stdoutFd), oldOutMode)
```
Do you intend for this to close over `err`?

timothyb89 replied:
Whoops, nope!

zmb3 reviewed (commented):
I don't know the code base well enough yet to give you much of a review, but added a few notes.

timothyb89 committed:
Address a few review comments

awly commented on lib/client/tncon/tncon.go:
```diff
@@ -0,0 +1,136 @@
+// ReadInputContinuous is a blocking call that continuously reads console
+// input events. Events will be emitted via channels to subscribers.
+func ReadInputContinuous() error {
+	runningMutex.Lock()
+	defer runningMutex.Unlock()
+
+	if running {
+		return fmt.Errorf("only one call to ReadInputContinuous is allowed")
+	}
```
`sync.Once` might be cleaner

awly commented on lib/client/session_windows.go:
```diff
@@ -45,22 +230,585 @@ func newSession(
+		events := tncon.Subscribe()
+		for event := range events {
+			switch ev := event.(type) {
+			case tncon.SequenceEvent:
```
what about other events?
if we don't care about them, maybe it's worth simplifying all the C code to just read sequences?

timothyb89 replied:
Resize events are useful but it was challenging to staple those into the resize handling I'd already written; I'll take another pass at them as I refactor.

I'd initially added mouse events expecting we'd need to implement that ourselves, but Windows is evidently translating those for us. I'll make sure they aren't needed and remove them (along with focus events).

timothyb89 replied:
I've now removed mouse and focus events.

awly commented on lib/client/tncon/tncon.c:
```diff
@@ -0,0 +1,883 @@
+{
+	// NOTE: small modification to send off an event here.
+	writeSequenceEvent(source, len);
+
+	while (len > 0) {
+		if (glob_outlen >= glob_space)
+			return glob_outlen;
+		*glob_out++ = *source++;
+		len--;
+		glob_outlen++;
+	}
+	return glob_outlen;
```
if we only care about our Go callback, can all this global buffer business be removed?

timothyb89 replied:
I've now removed the global buffer (and all other global state).

awly commented on lib/client/tncon/tncon.c:
```diff
@@ -0,0 +1,883 @@
+// This is a blocking call and will not return; callers should spawn this in
+// a background thread and subscribe to events via the Go interface.
+void ReadInputContinuous() {
+	HANDLE hInput = GetStdHandle(STD_INPUT_HANDLE);
+	char buf[32];
+
+	while (true) {
+		ReadConsoleForTermEmul(hInput, buf, 32);
+	}
```
maybe you can clone stdin handle and then close it to end this function?
having it run forever creates a lot of limitations

awly reviewed (commented):
Did an initial pass, want to understand the reasoning before going through all the C code.
Can you elaborate why exactly this forked C code is needed. What happens when you read from stdin in the windows terminal otherwise? Are there any redundant C parts or is it all required?

Regarding `session.go/session_windows.go`, let's refactor it:
- extract terminal handling code into `lib/client/terminal/`
- on all OSs, the API of that package should be something like:
```go
type Terminal struct{...}
func New() (*Terminal, error)
func (t *Terminal) Resize(w, h) error
func (t *Terminal) Stdin() io.Reader
func (t *Terminal) Stdout() io.Writer
func (t *Terminal) Stderr() io.Writer
func (t *Terminal) Close() error
```
- and implemented by `terminal_unix.go/terminal_windows.go/terminal_other.go`
- then merge `session.go/session_windows.go` into a single OS-agnostic file

timothyb89 commented on tool/tsh/tsh.go:
```diff
@@ -504,7 +504,6 @@ func Run(args []string, opts ...cliOption) error {
 	// because they all use a terminal.
 	if runtime.GOOS == constants.WindowsOS {
-		ssh.Hidden()
 		join.Hidden()
```
It does, I'll unhide it.

timothyb89 commented:
> Did an initial pass, want to understand the reasoning before going through all the C code.
> Can you elaborate why exactly this forked C code is needed. What happens when you read from stdin in the windows terminal otherwise? Are there any redundant C parts or is it all required?

The Windows console has various input processing modes that can be disabled with the [`SetConsoleMode()`](https://docs.microsoft.com/en-us/windows/console/setconsolemode) function. With the `ENABLE_PROCESSED_INPUT` and `ENABLE_LINE_INPUT` flags disabled and `ENABLE_VIRTUAL_TERMINAL_INPUT` enabled, the regular `stdin` stream returns sane VT sequences for most keystrokes. Unfortunately a few key sequences are still captured by the terminal, e.g. Ctrl-Z which forcefully kills the running app.

The native [`ReadConsoleInput()`](https://docs.microsoft.com/en-us/windows/console/readconsoleinput) function _does_ capture effectively all input events (when combined with the above flags, at least).

As for whether or not some of the C code is redundant, a large chunk potentially is. `GetVTSeqFromKeyStroke` translates native [`INPUT_RECORD`s](https://docs.microsoft.com/en-us/windows/console/input-record-str) to VT sequences. Recent versions of Windows (Win 10 1809+) translate events to VT sequences automatically when `ENABLE_VIRTUAL_TERMINAL_INPUT` is set and `tncon.c` will skip calling `GetVTSeqFromKeyStroke`. At the moment this patch requires 1809+ for ConPTY anyway, so it's effectively dead code.

I tentatively kept that functionality untouched in the event we might want to add support for older Windows versions, but if we stick with 1809+ it can be easily removed.

> Regarding `session.go/session_windows.go`, let's refactor it:

Thanks! I like this design a lot, I'll work on refactoring.

xacrimon commented on go.mod:
```diff
@@ -91,9 +93,7 @@ require (
 	github.com/tstranex/u2f v0.0.0-20160508205855-eb799ce68da4
 	github.com/vulcand/predicate v1.1.0
-	github.com/xeipuuv/gojsonpointer v0.0.0-20151027082146-e0fe6f683076 // indirect
-	github.com/xeipuuv/gojsonreference v0.0.0-20150808065054-e02fc20de94c // indirect
-	github.com/xeipuuv/gojsonschema v0.0.0-20151204154511-3988ac14d6f6 // indirect
+	github.com/xeipuuv/gojsonschema v1.2.0 // indirect
```
This looks like some overdue `go vendor` updates to me.

timothyb89 replied:
Yeah, this got included automatically via `make update-vendor`. I expect to rebase this commit eventually, so it may disappear then.

timothyb89 committed:
Remove significant chunks of unnecessary tncon code.

Removes the global buffer, `GetVTSeqFromKeyStroke`, and several
ancillary headers and functions that aren't needed for our (current)
use-case. Also removes mouse and focus events.

timothyb89 commented:
Made a first cleanup pass today, focusing on the C side of things. I removed the backwards-compat bits (`GetVTSeqFromKeyStroke`), removed all global state, and several unnecessary functions and headers.

awly commented:
Sounds great, let's support whatever versions Microsoft officially supports and remove all the legacy stuff.
If you're unsure about the rest of the C code and what to remove, we can do a pair programming session.

Ideally, we should understand what all the C code in our repo does and remove as much of it as possible. Any C code is a liability.

timothyb89 commented on lib/client/session_windows.go:
```diff
@@ -45,22 +230,585 @@ func newSession(
+		stdin = cpty.InPipe()
+	}
+	if stdout == nil {
+		stdin = cpty.OutPipe()
```
On this week's episode of "How Did This Ever Work?", I learn that conpty is entirely unnecessary for client-side SSH, and only noticed after realizing that this code has been super broken the whole time because of a typo.

... I'll be removing the conpty library along with the refactoring work. Will need to retest some things since we may now be compatible out of the box with older Win10 releases.

zmb3 commented on lib/client/api.go:
```diff
@@ -2786,7 +2786,18 @@ func Username() (string, error) {
+	username := u.Username
+
+	// If on Windows, strip the domain name.
+	if runtime.GOOS == "windows" {
```
```suggestion
	if runtime.GOOS == constants.WindowsOS {
```

zmb3 commented on lib/client/api.go:
```diff
@@ -2786,7 +2786,18 @@ func Username() (string, error) {
+	// If on Windows, strip the domain name.
+	if runtime.GOOS == "windows" {
+		idx := strings.LastIndex(username, "\\")
+		if idx > -1 {
```
I would check the upper bound too - unlikely but we don't want to panic with out of bounds error if the `\` is the last character.

```suggestion
		if idx > -1 && index < len(username) -1 {
```

timothyb89 committed:
Refactor OS-specific terminal handling

This significantly simplifies OS-specific terminal behavior:
 * Move OS specific terminal code into a new `terminal` package
 * Remove `session_windows.go` in favor of an OS-independent
   `session.go`, defer to terminal package for OS specific
   functionality.
 * Remove ConPTY since it's not needed.
 * Always wait for the terminal and ssh session to fully close before
   quitting.
 * Refactor tncon; ensure the raw reader can be closed and reopened,
   remove lots of unnecessary C code.

timothyb89 committed:
Revert dependency changes

timothyb89 committed:
Use WindowsOS constant.

timothyb89 committed:
Fix `tsh play` on Windows

This fixes `tsh play` on Windows by using the new `terminal` module to
initialize the terminal for raw input in a cross-platform way.

Additionally, this simplifies `terminal.New()` since in practice we never
want interactive mode at init time, and  fixes a broken unit test.

timothyb89 committed:
Use correct log library

timothyb89 committed:
Fix `tsh play` player controls on Windows

This fixes the console player controls on Windows as well as the timestamp
writer.

timothyb89 committed:
Clean up lints

timothyb89 committed:
Add missing license header

timothyb89 committed:
Fix broken unit test

timothyb89 commented:
I think this should be ready for review. A short summary of changes since the last time:

- Refactored `session.go` and `session_windows.go` into a single cross-platform `session.go`
- Added a new `terminal` package to manage OS-specific terminal management
- Removed conpty entirely
- Removed unnecessary backwards compatibility-related code from `tncon.c`
- Exposed `tsh ssh`, `tsh join`, `tsh scp`, `tsh join` on Windows
- `tncon` library can be closed and reopened as desired
- Fixed `tsh play`

timothyb89 commented:
I was able to test this on a fresh Windows 10 1607 VM and it looks like we support at least that far back now that we've dropped the ConPTY dependency (though admittedly mouse and resize support is broken - their ANSI emulation has improved since 2016).

timothyb89 committed:
Merge branch 'master' into timothyb89/tsh-ssh-windows

xacrimon reviewed (approved):
lgtm

timothyb89 committed:
Fix cross-compile builds on Linux/Docker

We need windows.h, which is not capitalized in the mingw packages
(and is case insensitive on Windows).

zmb3 commented on lib/client/terminal/terminal_common.go:
```diff
@@ -0,0 +1,60 @@
+	subscribersMutex sync.Mutex
+}
+
+// ResizeSubscribe creates a channel that will receive events whenever the
```
```suggestion
// Subscribe creates a channel that will receive events whenever the
```

zmb3 replied:
Seems like this is more general than just resize events. Do we need to update the godoc?

timothyb89 replied:
Ah, forgot to update that comment, will do so now.

zmb3 commented on lib/client/terminal/terminal_unix.go:
```diff
@@ -0,0 +1,160 @@
+	// Put the terminal into raw mode.
+	ts, err := term.SetRawTerminal(0)
+	if err != nil {
+		log.Warnf("Could not put terminal into raw mode: %v", err)
```
This is not an error? What happens if the terminal doesn't go into raw mode?

timothyb89 replied:
Without raw mode basic functionality can still (barely) work, so long as you don't try to use vim or something. On Unix in particular this was already a warning so we also don't want to add breaking behavior changes.

zmb3 commented on lib/client/terminal/terminal_unix.go:
```diff
@@ -0,0 +1,160 @@
+	signal.Notify(ctrlZSignal, syscall.SIGTSTP)
+	resizeSignal := make(chan os.Signal, 1)
+	signal.Notify(resizeSignal, syscall.SIGWINCH)
+	go func() {
```
This routine never terminates - is that okay?

timothyb89 replied:
In practice yes, but I'll fix it regardless.

zmb3 commented on lib/client/terminal/terminal_unix.go:
```diff
@@ -0,0 +1,160 @@
+// Resize makes a best-effort attempt to resize the terminal window. Support
+// varies between platforms and terminal emulators.
+func (t *Terminal) Resize(width, height int16) error {
+	os.Stdout.Write([]byte(fmt.Sprintf("\x1b[8;%d;%dt", height, width)))
```
Handle the Write error?

zmb3 commented on lib/client/terminal/terminal_unix.go:
```diff
@@ -0,0 +1,160 @@
+
+// Close closes the Terminal, restoring the console to its original state.
+func (t *Terminal) Close() error {
+	_ = t.closer.Close()
```
Why discard the error?

timothyb89 replied:
It could never return an error previously, but I've now just bubbled it up.

zmb3 commented on lib/client/terminal/terminal_windows.go:
```diff
@@ -0,0 +1,279 @@
+//go:build windows
+// +build windows
```
Shouldn't need the +build line anymore, right?

zmb3 replied:
In fact, you shouldn't need any build tags at all since the file is named terminal_windows.go

timothyb89 replied:
You're right, we can remove them at least from `terminal_windows.go`. I'm not sure we've fully moved to Go 1.17 everywhere, though, so it might be best to keep `+build` in `terminal_unix.go` for now?

zmb3 replied:
Gotcha, makes sense 👍

zmb3 commented on lib/client/terminal/terminal_windows.go:
```diff
@@ -0,0 +1,279 @@
+
+	oldOutMode, err := winterm.GetConsoleMode(uintptr(stdoutFd))
+	if err != nil {
+		return func() {}, fmt.Errorf("failed to retrieve stdout mode: %w", err)
```
I don't have a strong preference, but you can return nil instead of a nop func.

timothyb89 replied:
nil returns seem to generate nil pointer dereferences (https://play.golang.org/p/8IFuyejGiVt) so the nop func seems like a nicer thing to hand back to callers.

zmb3 replied:
Well yes you can't call the nil func, but I assumed the caller would do an error check first and only defer the cleanup if err == nil. Either way is fine though.

zmb3 commented on lib/client/terminal/terminal_windows.go:
```diff
@@ -0,0 +1,279 @@
+
+		err = winterm.SetConsoleMode(uintptr(stdinFd), newInMode)
+		if err != nil {
+			return func() {}, fmt.Errorf("failed to set stdin mode: %w", err)
```
In this case, you've set the mode for stdout and are not returning a cleanup func to restore the old mode? Is this what you want?

I can see a client not bothering to call the cleanup func if an error is returned, so maybe you want to just try restoring the stdout mode here before returning the error.

timothyb89 replied:
Good catch, it now attempts to reset the output mode on error.

zmb3 commented on lib/client/tncon/tncon.c:
```diff
@@ -0,0 +1,164 @@
+// +build windows,cgo
+
+/*
+* Author: Ray Hayes <ray.hayes@microsoft.com>
```
We've modified these files, right? If so, it would be nice to indicate that somehow.

timothyb89 replied:
I've added a line:

    Modifications for use in Teleport Copyright 2021 Gravitational, Inc.

It's a bit of an odd situation so I'm not sure if we have some preferred syntax for things like this.

zmb3 commented on lib/client/tncon/tncon.go:
```diff
@@ -0,0 +1,172 @@
+//go:build windows && cgo
+// +build windows,cgo
```
I think you can remove the +build line.

timothyb89 commented on lib/client/session.go:
```diff
@@ -523,14 +524,17 @@ func (ns *NodeSession) runCommand(ctx context.Context, cmd []string, callback Sh
 // watchSignals register UNIX signal handlers and properly terminates a remote shell session
 // must be called as a goroutine right after a remote shell is created
 func (ns *NodeSession) watchSignals(shell io.Writer) {
```
In practice no, but I'll clean this up anyway.

timothyb89 commented on lib/client/session.go:
```diff
@@ -245,47 +252,48 @@ func (ns *NodeSession) interactiveSession(callback interactiveCallback) error {
+	if ns.terminal.IsAttached() {
+		// Put the terminal into raw mode. Note that this must be done before
+		// pipeInOut() as it may replace streams.
+		ns.terminal.InitInteractive(true)
```
They are the same, but I'll rename this to `InitRaw` for clarity. The boolean argument controls whether or not raw input capture is enabled (only meaningful on Windows).

timothyb89 committed:
Address code review feedback

 - Rename `Terminal.InitInteractive` to `Terminal.InitRaw`
 - Ensure goroutines terminate on close
 - Fix outdated godoc comments
 - Ensure Terminal event subscribers are cleared (and their channels
   are closed)
 - Ensure terminal output mode is reset on error in initTerminal
 - Bubble up errors in Terminal.Close()
 - Add author notice to tncon.c re: our changes
 - Add go-ansiterm as a direct dependency
 - Run `make update-vendor`

timothyb89 committed:
Merge remote-tracking branch 'origin/master' into timothyb89/tsh-ssh-windows

timothyb89 commented on lib/client/player.go:
```diff
@@ -133,7 +134,7 @@ func timestampFrame(message string) {
 
 	// move cursor to -10:0
-	esc(fmt.Sprintf("[%d;%df", 0, int(sz.Width)-len(message)))
+	esc(fmt.Sprintf("[%d;%df", 0, int(width)-len(message)))
```
It would be off and result in some graphical strangeness. That said, the same problem also exists with any ANSI escapes in the message, so we'd want to filter those out to be totally correct here.

That said, we control the message here, and it is only ever set to a timestamp. Perhaps fixing this bug is out of scope for this PR?

zmb3 replied:
Ah yes I guess we're okay if the message is always a timestamp that we control.

timothyb89 replied:
I've added a TODO note about the issue to be safe

timothyb89 commented on lib/client/session.go:
```diff
@@ -542,15 +546,17 @@ func (ns *NodeSession) watchSignals(shell io.Writer) {
-				log.Errorf(err.Error())
+		for event := range events {
+			if _, ok := event.(terminal.StopEvent); ok {
+				_, err := shell.Write([]byte{26})
```
It is the control code for ^Z. To be honest, I don't totally know what a few of these are (0x03? I guess ETX / ^C?). At any rate, I'll pull them into constants.

timothyb89 commented:
`make update-vendor` seems _really_ intent on updating `gojsonschema` even though (as far as I can tell) this patch has not modified it. I can't clear up build errors without updating go modules (since go-ansiterm is now explicitly required) so... I guess that's included now?

timothyb89 committed:
Add constants and a small player.go TODO.

timothyb89 committed:
Clear linter warning

russjones reviewed (approved):
Bot.

timothyb89 committed:
Merge branch 'master' into timothyb89/tsh-ssh-windows
